The positions of introns are generally inferred from aligning spliced RNA to the genome using split-read mapping.  Determining the positions of introns in this manner requires that spliced RNA-Seq reads, when mapped to the genome, can be assigned uniquely. In this paper, we show that for the majority of annotated human splice sites (88.67%), this assumption does not hold. Even after accounting for the ambiguity of the AGgt and agGT splice signals at exon boundaries, 79.74% of introns are still ambiguously defined. This implies that without appealing to annotations or assumptions about intronic sequence structure, it is not possible to discover the genomic positions of splicing for at least 1 intron in 33,035 of 35,259 human genes using mRNA sequence alone.

See binary files attached to this release for analysis of human and mouse transcriptoms, as well as human annotated circular RNA and simulated transcriptomes. These files include equivalence junction sequences for annotated junctions as well as a summary of findings.
